packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx

```diff
@@ -211,29 +232,58 @@ const Root = React.forwardRef((props, forwardedRef) => {
 // ----------------------------------------------------------------------------
 
 /**
- * Returns theme from either parent context or by reading the closest
+ * Returns theme information from either parent ThemeContext or by reading the closest
  * data-iui-theme attribute if context is not found.
+ *
+ * Also returns the ThemeContext itself (if found).
  */
-const useParentTheme = () => {
+const useParentThemeAndContext = (rootElement: HTMLElement | null) => {
   const parentContext = React.useContext(ThemeContext);
-  const rootRef = React.useRef<HTMLElement>(null);
   const [parentThemeState, setParentTheme] = React.useState(
     parentContext?.theme,
   );
+  const [parentHighContrastState, setParentHighContrastState] = React.useState(
+    parentContext?.themeOptions?.highContrast,
+  );
+
+  const parentThemeRef = useLatestRef(parentContext?.theme);
 
   useIsomorphicLayoutEffect(() => {
-    setParentTheme(
-      (old) =>
-        old ||
-        (rootRef.current?.parentElement
-          ?.closest('[data-iui-theme]')
-          ?.getAttribute('data-iui-theme') as ThemeType),
+    // bail if we already have theme from context
```

Comment:
nit: clarity

```suggestion
    // bail if we already have theme (& highContrast) from context
```

Reply:
this is wrong, we don't care about highContrast and are not even checking for it

Reply:
To confirm, is this because if the theme is there from the context, then the highContrast would also be there? So just checking for theme is enough?

Reply:
yes and no. `theme` is definitely the important part, but `themeOptions.highContrast` can be undefined even when `theme` is present.